File: src_fortran/sub/vhs12.f
c
c
c
c
c
c
      subroutine VHS12  (MODE,LP,L1,M,U,INCU,UP,C,INCC,ICV,NCV)
C
      implicit none
C                                  SPECIFICATIONS FOR ARGUMENTS
      INTEGER            MODE,LP,L1,M,INCU,INCC,ICV,NCV
cuk      real               U(1),UP,C(1)
cuk      real               U(m+1),UP,C(ncv*m)  ! did NOT work...
      real               U(100),UP,C(100)   ! in this case only!!! (single-ev-l)
C                                  SPECIFICATIONS FOR LOCAL VARIABLES
      INTEGER            IJ,ILP,IL1,IM,INCR,I2,I3,I4,J
      real               SM,B
      real               ONE,CL,CLINV,SM1
C                                  FIRST EXECUTABLE STATEMENT
      ONE=1.e0
C
      IF (0.GE.LP.OR.LP.GE.L1.OR.L1.GT.M) GO TO 9005
      ILP=(LP-1)*INCU+1
      IL1=(L1-1)*INCU+1
      IM=(M-1)*INCU+1
      CL=ABS(U(ILP))
      IF (MODE.EQ.2) GO TO 15
C                                  CONSTRUCT THE TRANSFORMATION.
      DO 5 IJ=IL1,IM,INCU
    5 CL=MAX1(ABS(U(IJ)),CL)
      IF (CL.LE.0.0e0) GO TO 9005
      CLINV=ONE/CL
      SM=(U(ILP)*CLINV)**2
      DO 10 IJ=IL1,IM,INCU
   10 SM=SM+(U(IJ)*CLINV)**2
C                                  CONVERT DBLE. PREC. SM TO SNGL.
C                                    PREC. SM1
      SM1=SM
      CL=CL*SQRT(SM1)
      IF (U(ILP).GT.0.0e0) CL=-CL
      UP=U(ILP)-CL
      U(ILP)=CL
      GO TO 20
C                                  APPLY THE TRANSFORMATION
C                                    I+U*(U**T)/B TO C.
   15 IF (CL.LE.0.0e0) GO TO 9005
   20 IF (NCV.LE.0) GO TO 9005
      B=UP*U(ILP)
C                                  B MUST BE NONPOSITIVE HERE. IF B =
C                                    0., RETURN.
      IF (B.GE.0.0e0) GO TO 9005
      B=ONE/B
      I2=1-ICV+INCC*(LP-1)
      INCR=INCC*(L1-LP)
      DO 35 J=1,NCV
         I2=I2+ICV
         I3=I2+INCR
         I4=I3
         SM=C(I2)*UP
         DO 25 IJ=IL1,IM,INCU
            SM=SM+C(I3)*U(IJ)
            I3=I3+INCC
   25    CONTINUE
         IF (SM.EQ.0.0e0) GO TO 35
         SM=SM*B
         C(I2)=C(I2)+SM*UP
         DO 30 IJ=IL1,IM,INCU
            C(I4)=C(I4)+SM*U(IJ)
            I4=I4+INCC
   30    CONTINUE
   35 CONTINUE
 9005 RETURN
      END

File: src/sub/vhs12.c
/*
 * Householder matrix transformation (IMSL VHS12)
 * Applies orthogonal Householder transformation to a matrix
 * from Fortran VHS12 subroutine
 */

#include <stddef.h>
#include <stdio.h>
#include <math.h>

void vhs12(int mode, int lp, int l1, int m, double u[], double *up,
           double c[], int incu, int incc, int icv, int ncv) {
    int ij, ilp, il1, im, incr, i2, i3, i4, j;
    double sm, b;
    double one = 1.0f;
    double cl, clinv, sm1;
    
    /* Validate input parameters */
    if (lp <= 0 || lp >= l1 || l1 > m) {
        fprintf(stderr, "vhs12: invalid parameters or insufficient array sizes\n");
        return;
    }
    
    /* Calculate indices into arrays */
    ilp = (lp - 1) * incu;      /* 0-based index for LP element */
    il1 = (l1 - 1) * incu;      /* 0-based index for L1 element */
    im = (m - 1) * incu;        /* 0-based index for M element */
    
    cl = fabs(u[ilp]);
    
    if (mode != 2) {
        /* CONSTRUCT THE TRANSFORMATION */
        for (ij = il1; ij <= im; ij += incu) {
            cl = fmaxf(fabs(u[ij]), cl);
        }
        
        if (cl <= 0.0) {
            return;
        }
        
        clinv = one / cl;
        sm = (u[ilp] * clinv) * (u[ilp] * clinv);
        
        for (ij = il1; ij <= im; ij += incu) {
            sm = sm + (u[ij] * clinv) * (u[ij] * clinv);
        }
        
        sm1 = sm;  /* Convert to single precision */
        cl = cl * sqrtf(sm1);
        
        if (u[ilp] > 0.0) {
            cl = -cl;
        }
        
        *up = u[ilp] - cl;
        u[ilp] = cl;
    }
    
    /* APPLY THE TRANSFORMATION */
    if (ncv <= 0) {
        return;
    }
    
    if (cl <= 0.0) {
        return;
    }
    
    b = (*up) * u[ilp];
    
    /* B must be non-positive; if B == 0, return */
    if (b >= 0.0) {
        return;
    }
    
    b = one / b;
    
    /* Apply transformation I + U*(U^T)/B to C */
    i2 = 1 - icv + incc * (lp - 1);
    incr = incc * (l1 - lp);
    
    for (j = 0; j < ncv; j++) {
        i2 = i2 + icv;
        i3 = i2 + incr;
        i4 = i3;
        
        sm = c[i2] * (*up);
        
        // SEGFAULT here
        fprintf(stderr, "DEBUG: il1=%d, im=%d, incu=%d, len_c=%ld, len_u=%ld\n", il1, im, incu, (long)(sizeof(c)/sizeof(c[0])), (long)(sizeof(u)/sizeof(u[0])));
        for (ij = il1; ij <= im; ij += incu) {
            sm = sm + c[i3] * u[ij];
            i3 = i3 + incc;
        }
        
        if (sm == 0.0) {
            continue;
        }
        
        sm = sm * b;
        c[i2] = c[i2] + sm * (*up);
        
        for (ij = il1; ij <= im; ij += incu) {
            c[i4] = c[i4] + sm * u[ij];
            i4 = i4 + incc;
        }
    }
}
